File: tobas_core/tobas_rqt_tools/src/tobas_rqt_tools/widgets/spin_box.py
from overrides import override
from PyQt5.QtCore import QTimer
from PyQt5.QtWidgets import QWidget, QSpinBox, QDoubleSpinBox
from PyQt5.QtGui import QWheelEvent


class SpinBox(QSpinBox):
    """
    ===== QSpinBoxとの違い =====
    - 最大最小のデフォルト値をint32の最大最小に設定
    - マウスホイールイベントを無効化
    - フォーカス時にテキスト全体を選択
    """

    INT32_MAX = (1 << 31) - 1
    INT32_MIN = -(1 << 31)

    def __init__(self, parent: QWidget = None) -> None:
        super().__init__(parent)

        self.setMaximum(self.INT32_MAX)
        self.setMinimum(self.INT32_MIN)

    @override
    def wheelEvent(self, e: QWheelEvent) -> None:
        e.ignore()

    @override
    def focusInEvent(self, e) -> None:
        super().focusInEvent(e)
        QTimer.singleShot(0, lambda: self.selectAll())


class DoubleSpinBox(QDoubleSpinBox):
    """
    ===== QDoubleSpinBoxとの違い =====
    - 最大最小のデフォルト値を無限大に設定
    - マウスホイールイベントを無効化
    - フォーカス時にテキスト全体を選択
    """

    def __init__(self, parent: QWidget = None) -> None:
        super().__init__(parent)

        self.setMaximum(float("inf"))
        self.setMinimum(float("-inf"))

    @override
    def wheelEvent(self, e: QWheelEvent) -> None:
        e.ignore()

    @override
    def focusInEvent(self, e) -> None:
        super().focusInEvent(e)
        QTimer.singleShot(0, lambda: self.selectAll())

File: tobas_gui/tobas_qt_tools/src/widgets/spin_box.cpp
#include "tobas_qt_tools/widgets/spin_box.hpp"

#include <QTimer>
#include <QWheelEvent>

namespace qt
{
SpinBox::SpinBox(QWidget* parent) : super(parent)
{
  setMinimum(std::numeric_limits<int>::lowest());
  setMaximum(std::numeric_limits<int>::max());
}

void SpinBox::wheelEvent(QWheelEvent* event)
{
  event->ignore();
}

void SpinBox::focusInEvent(QFocusEvent* event)
{
  super::focusInEvent(event);
  QTimer::singleShot(0, this, &DoubleSpinBox::selectAll);
}

DoubleSpinBox::DoubleSpinBox(QWidget* parent) : super(parent)
{
  setMinimum(std::numeric_limits<double>::lowest());
  setMaximum(std::numeric_limits<double>::max());
}

void DoubleSpinBox::wheelEvent(QWheelEvent* event)
{
  event->ignore();
}

void DoubleSpinBox::focusInEvent(QFocusEvent* event)
{
  super::focusInEvent(event);
  QTimer::singleShot(0, this, &DoubleSpinBox::selectAll);
}
}  // namespace qt
